### PR TITLE
Bugfix: Handle repeated TEXT delimiter

### PR DESCRIPTION
**Pull request by cericson**
First, thanks for creating and maintaining this package! I use it a lot.

I discovered a bug in how the TEXT segment is parsed. A repeated delimiter is supposed to represent a single delimiter in the decoded strings (see item 3.2.7 of the [3.1 standard](http://software.broadinstitute.org/cancer/software/genepattern/attachments/fcs_3_1_standard.pdf)). For example, the key-value pair `{'flow_speed': '3 m/s'}` with the delimiter `/` should be encoded as:
`/flow_speed/3 m//s/`.
This string is currently decoded in fcsparser as:
`{'flow_speed': '3 m', '': 's'}`

The code in this PR should handle the delimiter correctly. It's not quite ready to merge yet, though. It appears that the test FCS file with duplicate column names actually just had the delimiter in the names, causing them to be read incorrectly. This means the corresponding test is now failing. If there are other real examples of FCS files with duplicate column names, we should swap one out. If not, perhaps the test can be deleted. What do you think?

A couple notes:
1) The same delimiter logic is described in the FCS 2.0 standard, so this change theoretically shouldn't cause any compatibility issues
2) I moved some of the surrounding code into a function just to make it more easily testable - the actual addition is only a few lines.

### CODE REVIEW

**Comment by eyurtsev** [comment]
Thanks will review during the weekend :)

File: fcsparser/api.py [hunk 1]
```diff
@@ -193,6 +193,31 @@ def read_header(self, file_handle, nextdata_offset=0):
...
+        if raw_text[-1] != delimiter:
+            raw_text = raw_text.strip()
+            if raw_text[-1] != delimiter:
+                msg = (u'The first two characters were:\n {}. The last two characters were: {}\n'
+                       u'Parser expects the same delimiter character in beginning '
+                       u'and end of TEXT segment'.format(raw_text[:2], raw_text[-2:]))
+                raise ParserFeatureNotImplementedError(msg)
+
+        # Below 1:-1 used to remove first and last characters which should be reserved for delimiter
+        # Also, the delimiter is "quoted" by being repeated (two consecutive delimiters)
```
Comment (eyurtsev) [inline fcsparser/api.py]: Is this part of the official fcs specification?

Could you paraphrase the comment to say that it is escaped by being repeated (rather than quoted)... that's the usual terminology for an escape character. 

---

Would you be able to rephrase this code to be more explicit / direct about what it is doing?

Instead of going through an intermediate array (`delimiter_split_arr`), perhaps we could directly identify the places to split? Any chance there's a simple regexp that could be used?

(the logic is quite difficult to follow)

File: fcsparser/api.py [hunk 2]
```diff
@@ -193,6 +193,31 @@ def read_header(self, file_handle, nextdata_offset=0):
...
+        delimiter_split_arr = [x.split(delimiter) for x in raw_text[1:-1].split(delimiter * 2)]
+
+        raw_text_segments = delimiter_split_arr[0]
+        for partial_segment in delimiter_split_arr[1:]:
+            raw_text_segments[-1] += (delimiter + partial_segment[0])
+            raw_text_segments.extend(partial_segment[1:])
+
+        keys, values = raw_text_segments[0::2], raw_text_segments[1::2]
+        text = {key: value for key, value in zip(keys, values)}  # build dictionary
```
Comment (eyurtsev) [inline fcsparser/api.py]: can replace with `text = dict(zip(keys,values))`

File: fcsparser/api.py (hunk 1 above)
Comment (cericson) [reply fcsparser/api.py]: Yes, it's been in since at least FCS 2.0 (see the links in the PR). They use the term "quoted", but I agree that "escaped" is more appropriate.

I was able to come up with a regexp which makes the process arguably more straightforward. I think it's still a little hairy, though. It needs both a look-ahead and a look-behind, which unfortunately prevent using `re.split()` (see the code comments I added).

File: fcsparser/api.py (hunk 2 above)
Comment (cericson) [reply fcsparser/api.py]: Good call.

**Comment by cericson** [comment]
Have you found any other examples of FCS files with duplicate feature names?

File: fcsparser/api.py [hunk 3]
```diff
@@ -206,16 +207,23 @@ def _extract_text_dict(self, raw_text):
...
-        for partial_segment in delimiter_split_arr[1:]:
-            raw_text_segments[-1] += (delimiter + partial_segment[0])
-            raw_text_segments.extend(partial_segment[1:])
+        raw_text = raw_text[1:-1]
+
+        ed = re.escape(delimiter)  # escaped for regex, not FCS format
+        # The delimiter is escaped by being repeated (two consecutive delimiters). This regex finds
+        # odd-length repetitions of the delimiter. Since each pair is escaped, only the last
+        # delimiter in odd-length repetitions truly delimits TEXT keys and values.
+        regex = "(?<!" + ed + ")(" + ed * 2 + ")*" + ed + "(?!" + ed + ")"
```
Comment (eyurtsev) [inline fcsparser/api.py]: Since we're allowing the test to construct the regular expression dynamically, we should check for a few edge cases with the delimiter being `\` rather than `/` 

You may nee to specify that the regexp is a raw string `regex = r'...'`

File: fcsparser/api.py [hunk 4]
```diff
@@ -206,16 +207,23 @@ def _extract_text_dict(self, raw_text):
...
+        # We can't directly split with this regexp because the first n-1 delimiters would have to
+        # be in the look-behind, and look-behinds can only match fixed-length patterns.
+        split_idx = [0] + [m.end() for m in re.finditer(regex, raw_text)] + [len(raw_text) + 1]
+        # Use the identified split indices to actually split the TEXT segment, and replace the
+        # escaped delimiters along the way
+        raw_text_segments = [raw_text[start:(end - 1)].replace('//', '/')
```
Comment (eyurtsev) [inline fcsparser/api.py]: The delimiter doesn't have to `/`

File: fcsparser/api.py (hunk 3 above)
Comment (eyurtsev) [inline fcsparser/api.py]: Let me think about this a bit tonight to see if I can come up with anything different (the previous version you had is actually easier to read)

File: fcsparser/api.py (hunk 4 above)
Comment (cericson) [reply fcsparser/api.py]: Good point, I missed that.

File: fcsparser/api.py (hunk 3 above)
Comment (cericson) [reply fcsparser/api.py]: At least one of the test FCS files has `\` as a delimiter. I had to add the `re.escape()` to handle that.

Comment (cericson) [reply fcsparser/api.py]: For what it's worth, I had to go through a few different iterations of the code to get to the previous version. Maybe I revert to that, and add some better comments?

Comment (eyurtsev) [reply fcsparser/api.py]: I apologize I forgot to look at it -- but yes that sounds good

Comment (eyurtsev) [reply fcsparser/api.py]: Sorry for making you go through the extra trouble!

Comment (cericson) [reply fcsparser/api.py]: No problem - it got me to finally learn how lookaheads work.

**Comment by cericson** [comment]
Unless you object, I'll delete the failing test since its assumptions don't appear to be valid any more. The FCS spec ($PnN subsection of 3.2.20) forbids duplicate parameter names.

File: fcsparser/api.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
 import string
 import sys
 import warnings
+import re
```
Comment (eyurtsev) [inline fcsparser/api.py]: FYI The imports should be alphabetically sorted (PEP-8) -- i'll fix this up after the merge

Comment (cericson) [reply fcsparser/api.py]: Oh, good to know.
